In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('ga_hits.csv')
df.head(100)

session_id    hit_date  hit_time  \
0   5639623078712724064.1640254056.1640254056  2021-12-23  597864.0   
1   7750352294969115059.1640271109.1640271109  2021-12-23  597331.0   
2    885342191847998240.1640235807.1640235807  2021-12-23  796252.0   
3    142526202120934167.1640211014.1640211014  2021-12-23  934292.0   
4   3450086108837475701.1640265078.1640265078  2021-12-23  768741.0   
..                                        ...         ...       ...   
95  8605348810358797286.1640248293.1640248293  2021-12-23     396.0   
96  8652515561382901493.1640265465.1640265465  2021-12-23    1257.0   
97  8456784096293865033.1640285770.1640285770  2021-12-23      90.0   
98  8467913524788830717.1640210941.1640210941  2021-12-23      28.0   
99  8889851850682284318.1640246080.1640246080  2021-12-23    1131.0   

    hit_number hit_type           hit_referer  \
0           30    event                   NaN   
1           41    event                   NaN   
2           49    event                   NaN   
3           46    event                   NaN   
4           79    event                   NaN   
..         ...      ...                   ...   
95           7    event  trXmoxcphNjuYcShOXwL   
96           7    event                   NaN   
97           4    event                   NaN   
98           3    event  InufokblhLdayYnloSPA   
99           5    event                   NaN   

                                        hit_page_path event_category  \
0   sberauto.com/cars?utm_source_initial=google&ut...           quiz   
1   sberauto.com/cars/fiat?city=1&city=18&rental_c...           quiz   
2   sberauto.com/cars/all/volkswagen/polo/e994838f...           quiz   
3   sberauto.com/cars?utm_source_initial=yandex&ut...           quiz   
4   sberauto.com/cars/all/mercedes-benz/cla-klasse...           quiz   
..                                                ...            ...   
95  sberauto.com/cars/all/volkswagen/polo/e994838f...       card_web   
96  sberauto.com/cars/all/mercedes-benz/e-klasse/d...       card_web   
97  sberauto.com/cars/all/renault/logan/8c3c73f2?r...       card_web   
98  sberauto.com/cars/all/renault/duster/145d53f1?...       card_web   
99  sberauto.com/cars/all/skoda/rapid/bf24b977?ren...       card_web   

   event_action event_label  event_value  
0     quiz_show         NaN          NaN  
1     quiz_show         NaN          NaN  
2     quiz_show         NaN          NaN  
3     quiz_show         NaN          NaN  
4     quiz_show         NaN          NaN  
..          ...         ...          ...  
95    view_card         NaN          NaN  
96    view_card         NaN          NaN  
97    view_card         NaN          NaN  
98    view_card         NaN          NaN  
99    view_card         NaN          NaN  

[100 rows x 11 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15726470 entries, 0 to 15726469
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   session_id      object 
 1   hit_date        object 
 2   hit_time        float64
 3   hit_number      int64  
 4   hit_type        object 
 5   hit_referer     object 
 6   hit_page_path   object 
 7   event_category  object 
 8   event_action    object 
 9   event_label     object 
 10  event_value     float64
dtypes: float64(2), int64(1), object(8)
memory usage: 1.3+ GB


In [4]:
df.describe(include='all')

session_id    hit_date      hit_time  \
count                                    15726470    15726470  6.566148e+06   
unique                                    1734610         226           NaN   
top     5442565791571325612.1632449195.1632449195  2021-05-24           NaN   
freq                                          768      513035           NaN   
mean                                          NaN         NaN  2.091050e+05   
std                                           NaN         NaN  4.032110e+05   
min                                           NaN         NaN  0.000000e+00   
25%                                           NaN         NaN  2.412000e+04   
50%                                           NaN         NaN  8.994100e+04   
75%                                           NaN         NaN  2.195352e+05   
max                                           NaN         NaN  1.315688e+07   

          hit_number  hit_type           hit_referer           hit_page_path  \
count   1.572647e+07  15726470               9451666                15726470   
unique           NaN         1                 37873                  342715   
top              NaN     event  HbolMJUevblAbkHClEQa  podpiska.sberauto.com/   
freq             NaN  15726470               8879187                 2793639   
mean    2.356715e+01       NaN                   NaN                     NaN   
std     2.887713e+01       NaN                   NaN                     NaN   
min     1.000000e+00       NaN                   NaN                     NaN   
25%     7.000000e+00       NaN                   NaN                     NaN   
50%     1.500000e+01       NaN                   NaN                     NaN   
75%     2.900000e+01       NaN                   NaN                     NaN   
max     5.000000e+02       NaN                   NaN                     NaN   

       event_category event_action           event_label  event_value  
count        15726470     15726470              11966286          0.0  
unique             52          230                 39825          NaN  
top          card_web    view_card  KclpemfoHstknWHFiLit          NaN  
freq          7456998      3558985               6505447          NaN  
mean              NaN          NaN                   NaN          NaN  
std               NaN          NaN                   NaN          NaN  
min               NaN          NaN                   NaN          NaN  
25%               NaN          NaN                   NaN          NaN  
50%               NaN          NaN                   NaN          NaN  
75%               NaN          NaN                   NaN          NaN  
max               NaN          NaN                   NaN          NaN

In [5]:
df2 = pd.read_csv('ga_sessions.csv', low_memory=False)
df2.head(100)

session_id              client_id  \
0   9055434745589932991.1637753792.1637753792  2108382700.1637753791   
1    905544597018549464.1636867290.1636867290   210838531.1636867288   
2   9055446045651783499.1640648526.1640648526  2108385331.1640648523   
3   9055447046360770272.1622255328.1622255328  2108385564.1622255328   
4   9055447046360770272.1622255345.1622255345  2108385564.1622255328   
..                                        ...                    ...   
95  9055856575798829429.1633783120.1633783120  2108480915.1633673589   
96  9055856575798829429.1633783182.1633783182  2108480915.1633673589   
97  9055857473438318961.1624998265.1624998265  2108481124.1624998257   
98     9055858569671852.1630434476.1630434476     2108481.1630434476   
99  9055865509330223519.1633091999.1633091999  2108482995.1633091999   

    visit_date visit_time  visit_number            utm_source  \
0   2021-11-24   14:36:32             1  ZpYIoDJMcFzVoPFsHGJL   
1   2021-11-14   08:21:30             1  MvfHsxITijuriZxsqZqt   
2   2021-12-28   02:42:06             1  ZpYIoDJMcFzVoPFsHGJL   
3   2021-05-29   05:00:00             1  kjsLglQLzykiRbcDiGcD   
4   2021-05-29   05:00:00             2  kjsLglQLzykiRbcDiGcD   
..         ...        ...           ...                   ...   
95  2021-10-09   15:00:00             3  ZpYIoDJMcFzVoPFsHGJL   
96  2021-10-09   15:00:00             4  ZpYIoDJMcFzVoPFsHGJL   
97  2021-06-29   23:00:00             1  bByPQxmDaMXgpHeypKSM   
98  2021-08-31   21:00:00             1  ISrKoXQCxqqYvAZICvjs   
99  2021-10-01   15:00:00             1  kjsLglQLzykiRbcDiGcD   

         utm_medium          utm_campaign         utm_adcontent  \
0            banner  LEoPHuyFvzoNfnzGgfcd  vCIpmpaGBnIQhyYNkXqp   
1               cpm  FTjNLDyTrXaWYgZymFkV  xhoenQgDQsgfEPYNPwKO   
2            banner  LEoPHuyFvzoNfnzGgfcd  vCIpmpaGBnIQhyYNkXqp   
3               cpc                   NaN  NOBKLgtuvqYWkXQHeYWM   
4               cpc                   NaN                   NaN   
..              ...                   ...                   ...   
95           banner  LEoPHuyFvzoNfnzGgfcd  JNHcPlZPxEMWDnRiyoBf   
96           banner  SgIUDYUKnyWHVowUOqid  JNHcPlZPxEMWDnRiyoBf   
97         referral                   NaN                   NaN   
98  blogger_stories  BAZCuyHZnaPrMGOMrcCQ  JNHcPlZPxEMWDnRiyoBf   
99              cpc  nSReTmyFtbSjlPrTKoaX                   NaN   

             utm_keyword device_category device_os device_brand device_model  \
0   puhZPIYqKXeFPaUviSjo          mobile   Android       Huawei          NaN   
1   IGUCNvHlhfHpROGclCit          mobile   Android      Samsung          NaN   
2   puhZPIYqKXeFPaUviSjo          mobile   Android       Huawei          NaN   
3                    NaN          mobile       NaN       Xiaomi          NaN   
4                    NaN          mobile       NaN       Xiaomi          NaN   
..                   ...             ...       ...          ...          ...   
95                   NaN          mobile       NaN        Apple          NaN   
96                   NaN          mobile       NaN        Apple          NaN   
97                   NaN          mobile       NaN      Samsung          NaN   
98                   NaN          mobile       NaN        Apple          NaN   
99                   NaN         desktop       NaN          NaN          NaN   

   device_screen_resolution    device_browser geo_country     geo_city  
0                   360x720            Chrome      Russia     Zlatoust  
1                   385x854  Samsung Internet      Russia       Moscow  
2                   360x720            Chrome      Russia  Krasnoyarsk  
3                   393x786            Chrome      Russia       Moscow  
4                   393x786            Chrome      Russia       Moscow  
..                      ...               ...         ...          ...  
95                  390x844            Safari      Russia    Krasnodar  
96                  390x844            

In [6]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1860042 entries, 0 to 1860041
Data columns (total 18 columns):
 #   Column                    Dtype 
---  ------                    ----- 
 0   session_id                object
 1   client_id                 object
 2   visit_date                object
 3   visit_time                object
 4   visit_number              int64 
 5   utm_source                object
 6   utm_medium                object
 7   utm_campaign              object
 8   utm_adcontent             object
 9   utm_keyword               object
 10  device_category           object
 11  device_os                 object
 12  device_brand              object
 13  device_model              object
 14  device_screen_resolution  object
 15  device_browser            object
 16  geo_country               object
 17  geo_city                  object
dtypes: int64(1), object(17)
memory usage: 255.4+ MB


In [7]:
df2.describe(include='all')

session_id              client_id  \
count                                     1860042                1860042   
unique                                    1860042                1391719   
top     8656039981475191037.1634228491.1634228491  1750498477.1629450062   
freq                                            1                    462   
mean                                          NaN                    NaN   
std                                           NaN                    NaN   
min                                           NaN                    NaN   
25%                                           NaN                    NaN   
50%                                           NaN                    NaN   
75%                                           NaN                    NaN   
max                                           NaN                    NaN   

        visit_date visit_time  visit_number            utm_source utm_medium  \
count      1860042    1860042  1.860042e+06               1859945    1860042   
unique         226      85318           NaN                   293         56   
top     2021-05-24   12:00:00           NaN  ZpYIoDJMcFzVoPFsHGJL     banner   
freq         39453      61067           NaN                578290     552272   
mean           NaN        NaN  2.712804e+00                   NaN        NaN   
std            NaN        NaN  1.182907e+01                   NaN        NaN   
min            NaN        NaN  1.000000e+00                   NaN        NaN   
25%            NaN        NaN  1.000000e+00                   NaN        NaN   
50%            NaN        NaN  1.000000e+00                   NaN        NaN   
75%            NaN        NaN  2.000000e+00                   NaN        NaN   
max            NaN        NaN  5.640000e+02                   NaN        NaN   

                utm_campaign         utm_adcontent           utm_keyword  \
count                1640439               1524427                777981   
unique                   412                   286                  1219   
top     LTuZkdKfxRGVceoWkVyg  JNHcPlZPxEMWDnRiyoBf  puhZPIYqKXeFPaUviSjo   
freq                  463481               1006599                506819   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       device_category device_os device_brand          device_model  \
count          1860042    789904      1492864                 16338   
unique               3        13          205                   104   
top             mobile   Android        Apple  AuMdmADEIoPXiWpTsBEj   
freq           1474871    464054       551088                  9778   
mean               NaN       NaN          NaN                   NaN   
std                NaN       NaN          NaN                   NaN   
min                NaN       NaN          NaN                   NaN   
25%                NaN       NaN          NaN                   NaN   
50%                NaN       NaN          NaN                   NaN   
75%                NaN       NaN          NaN                   NaN   
max                NaN       NaN          NaN                   NaN   

       device_screen_resolution device_browser geo_country geo_city  
count                   1860042        1860042     1860042  1860042  
unique                     5039             57         166     2548  
top                     414x896         Chrome      Russia   Moscow  
freq                     169090        1013436     1800565   805329  
mean                        NaN            NaN   

In [8]:
df.duplicated().any()

False

In [9]:
df2.duplicated().any()

False

In [10]:
df['session_id'].unique()

array(['5639623078712724064.1640254056.1640254056',
       '7750352294969115059.1640271109.1640271109',
       '885342191847998240.1640235807.1640235807', ...,
       '3349670872968620291.1640264206.1640264206',
       '1009509786569589790.1640244938.1640244938',
       '5381267721977892188.1640271196.1640271196'], dtype=object)

In [11]:
df[df.event_action.isin(['sub_car_claim_click', 'sub_car_claim_submit_click', 'sub_open_dialog_click',
                         'sub_custom_question_submit_click', 'sub_call_number_click',
                         'sub_callback_submit_click', 'sub_submit_success', 'sub_car_request_submit_click'])]

session_id    hit_date  hit_time  \
4016      2744563715298057088.1640258436.1640258436  2021-12-23  843092.0   
4045      3087297479839089634.1640268774.1640268774  2021-12-23  194144.0   
4046      3156966333326004302.1640206419.1640206800  2021-12-23  327223.0   
4047      3750243879753098158.1640272208.1640272208  2021-12-23  156992.0   
4048      7518333712042258254.1640258901.1640258901  2021-12-23  170616.0   
...                                             ...         ...       ...   
15725025  1277864870843199549.1636773954.1636773954  2021-11-13  354017.0   
15725133   965861352491898977.1636751459.1636751459  2021-11-13  172590.0   
15725134  4727705172767122620.1636818847.1636818847  2021-11-13  356120.0   
15725135  4727705172767122620.1636818847.1636818847  2021-11-13  312303.0   
15725136   116562085450617659.1636832061.1636832061  2021-11-13  124047.0   

          hit_number hit_type hit_referer  \
4016              81    event         NaN   
4045              22    event         NaN   
4046              63    event         NaN   
4047              20    event         NaN   
4048              16    event         NaN   
...              ...      ...         ...   
15725025          30    event         NaN   
15725133          18    event         NaN   
15725134          43    event         NaN   
15725135          41    event         NaN   
15725136          26    event         NaN   

                                              hit_page_path    event_category  \
4016      sberauto.com/cars/all/kia/rio/fee33fe6?utm_sou...        sub_submit   
4045      sberauto.com/cars/all/skoda/rapid/bf24b977?utm...        sub_submit   
4046      sberauto.com/cars/all/skoda/rapid/bf24b977?utm...        sub_submit   
4047      sberauto.com/cars/all/nissan/x-trail/0744675f?...        sub_submit   
4048      sberauto.com/cars/all/mercedes-benz/gla-klasse...        sub_submit   
...                                                     ...               ...   
15725025  sberauto.com/cars/all/lada-vaz/vesta/2fc745ed?...  sub_button_click   
15725133  sberauto.com/cars/all/volkswagen/polo/e994838f...        sub_submit   
15725134  sberauto.com/cars/all/volkswagen/polo/e994838f...        sub_submit   
15725135  sberauto.com/cars/all/volkswagen/polo/e994838f...  sub_button_click   
15725136  sberauto.com/cars/all/volkswagen/polo/e994838f...  sub_button_click   

                        event_action           event_label  event_value  
4016              sub_submit_success  nsPPIRqjxBefONGPpnsF          NaN  
4045              sub_submit_success  nsPPIRqjxBefONGPpnsF          NaN  
4046              sub_submit_success  nsPPIRqjxBefONGPpnsF          NaN  
4047              sub_submit_success  nsPPIRqjxBefONGPpnsF          NaN  
4048              sub_submit_success  KuMiABMMbspIDDhiCNVS          NaN  
...                              ...                   ...          ...  
15725025  sub_car_claim_submit_click                   NaN          NaN  
15725133          sub_submit_success  nsPPIRqjxBefONGPpnsF          NaN  
15725134          sub_submit_success  uimgZZmhfLQwbKAZZfCk          NaN  
15725135       sub_open_dialog_click  ZaZuwAXOKlbzyhUqtnmk          NaN  
15725136  sub_car_claim_submit_click                   NaN          NaN  

[104908 rows x 11 columns]

In [12]:
df2[df2['session_id'] == '9055794766915104430.1624305327.1624305327']

session_id              client_id  \
78  9055794766915104430.1624305327.1624305327  2108466524.1624305326   

    visit_date visit_time  visit_number            utm_source utm_medium  \
78  2021-06-21   22:00:00             1  BHcvLfOaCWvWTykYqHVe        cpc   

   utm_campaign utm_adcontent utm_keyword device_category device_os  \
78          NaN           NaN         NaN         desktop       NaN   

   device_brand device_model device_screen_resolution device_browser  \
78          NaN          NaN                 834x1112         Safari   

   geo_country    geo_city  
78      Russia  Domodedovo

In [13]:
df[df['session_id'] == '9055794766915104430.1624305327.1624305327']

session_id    hit_date  hit_time  \
5593195  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5594864  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5594950  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5595467  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5605713  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5605736  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5606138  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5606173  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5621553  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5621559  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5621575  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5621883  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5621906  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5621910  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5622967  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5622979  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623008  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623408  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623447  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623608  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623928  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623995  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5623999  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624160  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624181  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624221  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624288  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624299  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624322  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624455  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624481  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5624483  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5626638  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5626654  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5626701  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5628464  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5628611  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5628878  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5630262  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5631623  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5631625  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5631888  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5631896  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5632516  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5633912  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5634067  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5634435  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5637684  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5637685  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5644414  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   
5644415  9055794766915104430.1624305327.1624305327  2021-06-21       NaN   

         hit_number hit_type           hit_referer  \
5593195           2    event  Hbo

In [14]:
df.isna().any()

session_id        False
hit_date          False
hit_time           True
hit_number        False
hit_type          False
hit_referer        True
hit_page_path     False
event_category    False
event_action      False
event_label        True
event_value        True
dtype: bool

In [15]:
df2.isna().any()

session_id                  False
client_id                   False
visit_date                  False
visit_time                  False
visit_number                False
utm_source                   True
utm_medium                  False
utm_campaign                 True
utm_adcontent                True
utm_keyword                  True
device_category             False
device_os                    True
device_brand                 True
device_model                 True
device_screen_resolution    False
device_browser              False
geo_country                 False
geo_city                    False
dtype: bool

# 1. Разведочный анализ.

1.1. Датасет представлен двумя таблицами. При этом первичной является таблица сессий, а вторичной таблица хитов.

1.2. В обеих таблицах отсутствуют дубли.

1.3. Исходя из количества уникальных строк сессий в обеих таблицах и общего количества строк хитов становится очевидно, что существуют сессии, как без единого хита, так и сессии где хитов 1 и более.

1.4. Количество уникальных клиентов говорит о том, что в период, который охватывает датасет существуют клиенты, которые посещали сайт не один раз.

1.5. Однако анализ данных показал, что в датасете отсутствуют представления о визитах начиная с первого для каждого из клиентов, соответственно невозможно учесть посещение клиентов с аккумулированием времени визита. Таким образом для модели будет считаться, что каждый визит уникален и независим друг от друга.

1.6. В дате визита отсутствуют пропуски.

1.7. Во времени визита отсутствуют пропуски и это время указывает нам на время начала сессии.

1.8. В порядковом номере визита отсутствуют пропуски, однако как уже было отмечено выше, нет цельного представления о каждом визите клиента начиная с первого.

1.9. Информация о канале привлечения содержит 97 пропусков. При 293 уникальных значений имеет смысл замены на наиболее часто встречающийся или наиболее часто встречающийся применительно к типу привлечения.

1.10. В типе привлечения отсутствуют пропуски.

1.11. Информация о рекламной компании содержит 219603 пропуска, что составляет около 12% от общего количества. Удаление этой фичи не видится целесообразным, однако например для значения  utm_source = BHcvLfOaCWvWTykYqHVe, что составляет около половины всех пропусков какие-либо данные отсутствуют, поэтому в рекламной компании все пропуски будем заменять на значение other.

1.12. Информация utm_adcontent содержит 335615 пропусков, что составляет около 18% от общего количества. Удаление этой фичи не видится целесообразным, однако например для значения  utm_source = BHcvLfOaCWvWTykYqHVe, что составляет около трети всех пропусков какие-либо данные отсутствуют, поэтому в utm_adcontent все пропуски будем заменять на значение other.

1.13. Информация utm_keyword содержит пропусков более 50% от общего количества. Фича на удаление.

1.14. Информация device_category не содержит пропусков.

1.15. Информация device_os содержит пропусков более 50% от общего количества. Фича на удаление.

1.16. Информация device_brand содержит 367178 пропусков, что составляет около 20% от общего количества. Удаление этой фичи не видится целесообразным, поэтому в device_brand все пропуски будем заменять на значение other.

1.17. Информация device_model подавляющее количество пропусков. Фича на удаление.

1.18. Информация device_screen_resolution, device_browser, geo_country, geo_city не содержат пропусков.

1.19. Информация hit_number, hit_page_path, event_category, event_action не содержат пропусков.

1.20. Информация hit_date не содержит пропусков, однако более 50% количество пропусков в hit_time делает нецелесообразным использование данного поля, при том, что время начала сессии все равно известно и находится в visit_time. Комплексно продолжительность сессии здесь будет учитываться за счет времени начала сессии и как количества так и порядкового номера нажатия в hit_number. Фича на удаление. 

1.21. Информация hit_time содержит пропусков более 50% от общего количества. Фича на удаление.

1.22. Информация hit_type не содержит пропусков, однако значения представлены единственным уникальным. Фича на удаление.

1.23. Информация hit_referer содержит 6274804 пропусков, что составляет около 40% от общего количества. Фича на удаление.

1.24. Информация event_label содержит 3760184 пропусков, что составляет около 24% от общего количества. Фича на удаление.

1.25. Информация event_value содержит только пропуски. Фича на удаление.

# 2. Подготовка данных.

2.1. Выполняем загрузку данных без фич отсеянных на этапе ознакомления с датасетом.

In [16]:
df_hits = pd.read_csv('ga_hits.csv', usecols = [
    'session_id', 'hit_number', 'hit_page_path', 'event_category', 'event_action'])

In [17]:
df_hits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15726470 entries, 0 to 15726469
Data columns (total 5 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   session_id      object
 1   hit_number      int64 
 2   hit_page_path   object
 3   event_category  object
 4   event_action    object
dtypes: int64(1), object(4)
memory usage: 599.9+ MB


In [18]:
df_hits.describe(include='all')

session_id    hit_number  \
count                                    15726470  1.572647e+07   
unique                                    1734610           NaN   
top     5442565791571325612.1632449195.1632449195           NaN   
freq                                          768           NaN   
mean                                          NaN  2.356715e+01   
std                                           NaN  2.887713e+01   
min                                           NaN  1.000000e+00   
25%                                           NaN  7.000000e+00   
50%                                           NaN  1.500000e+01   
75%                                           NaN  2.900000e+01   
max                                           NaN  5.000000e+02   

                 hit_page_path event_category event_action  
count                 15726470       15726470     15726470  
unique                  342715             52          230  
top     podpiska.sberauto.com/       card_web    view_card  
freq                   2793639        7456998      3558985  
mean                       NaN            NaN          NaN  
std                        NaN            NaN          NaN  
min                        NaN            NaN          NaN  
25%                        NaN            NaN          NaN  
50%                        NaN            NaN          NaN  
75%                        NaN            NaN          NaN  
max                        NaN            NaN          NaN

In [19]:
df_sessions = pd.read_csv('ga_sessions.csv', usecols = [
    'session_id', 'client_id', 'visit_date', 'visit_time', 'visit_number', 'utm_source', 'utm_medium', 'utm_campaign',
    'utm_adcontent', 'device_category', 'device_brand', 'device_screen_resolution', 'device_browser', 'geo_country',
    'geo_city'], low_memory=False)

In [20]:
df_sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1860042 entries, 0 to 1860041
Data columns (total 15 columns):
 #   Column                    Dtype 
---  ------                    ----- 
 0   session_id                object
 1   client_id                 object
 2   visit_date                object
 3   visit_time                object
 4   visit_number              int64 
 5   utm_source                object
 6   utm_medium                object
 7   utm_campaign              object
 8   utm_adcontent             object
 9   device_category           object
 10  device_brand              object
 11  device_screen_resolution  object
 12  device_browser            object
 13  geo_country               object
 14  geo_city                  object
dtypes: int64(1), object(14)
memory usage: 212.9+ MB


In [21]:
df_sessions.describe(include='all')

session_id              client_id  \
count                                     1860042                1860042   
unique                                    1860042                1391719   
top     8656039981475191037.1634228491.1634228491  1750498477.1629450062   
freq                                            1                    462   
mean                                          NaN                    NaN   
std                                           NaN                    NaN   
min                                           NaN                    NaN   
25%                                           NaN                    NaN   
50%                                           NaN                    NaN   
75%                                           NaN                    NaN   
max                                           NaN                    NaN   

        visit_date visit_time  visit_number            utm_source utm_medium  \
count      1860042    1860042  1.860042e+06               1859945    1860042   
unique         226      85318           NaN                   293         56   
top     2021-05-24   12:00:00           NaN  ZpYIoDJMcFzVoPFsHGJL     banner   
freq         39453      61067           NaN                578290     552272   
mean           NaN        NaN  2.712804e+00                   NaN        NaN   
std            NaN        NaN  1.182907e+01                   NaN        NaN   
min            NaN        NaN  1.000000e+00                   NaN        NaN   
25%            NaN        NaN  1.000000e+00                   NaN        NaN   
50%            NaN        NaN  1.000000e+00                   NaN        NaN   
75%            NaN        NaN  2.000000e+00                   NaN        NaN   
max            NaN        NaN  5.640000e+02                   NaN        NaN   

                utm_campaign         utm_adcontent device_category  \
count                1640439               1524427         1860042   
unique                   412                   286               3   
top     LTuZkdKfxRGVceoWkVyg  JNHcPlZPxEMWDnRiyoBf          mobile   
freq                  463481               1006599         1474871   
mean                     NaN                   NaN             NaN   
std                      NaN                   NaN             NaN   
min                      NaN                   NaN             NaN   
25%                      NaN                   NaN             NaN   
50%                      NaN                   NaN             NaN   
75%                      NaN                   NaN             NaN   
max                      NaN                   NaN             NaN   

       device_brand device_screen_resolution device_browser geo_country  \
count       1492864                  1860042        1860042     1860042   
unique          205                     5039             57         166   
top           Apple                  414x896         Chrome      Russia   
freq         551088                   169090        1013436     1800565   
mean            NaN                      NaN            NaN         NaN   
std             NaN                      NaN            NaN         NaN   
min             NaN                      NaN            NaN         NaN   
25%             NaN                      NaN            NaN         NaN   
50%             NaN                      NaN            NaN         NaN   
75%             NaN                      NaN            NaN         NaN   
max             NaN                      NaN            NaN         NaN   

       geo_city  
count   1860042  
unique     2548  
top      Moscow  
freq     805329  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

2.2. Заполняем пропуски и проверяем, что все пропуски заполнены.

In [22]:
df_sessions.utm_source.fillna('ZpYIoDJMcFzVoPFsHGJL', inplace=True)

In [23]:
df_sessions.utm_campaign.fillna('other', inplace=True)

In [24]:
df_sessions.utm_adcontent.fillna('other', inplace=True)

In [25]:
df_sessions.device_brand.fillna('other', inplace=True)

In [26]:
df_hits.isna().any()

session_id        False
hit_number        False
hit_page_path     False
event_category    False
event_action      False
dtype: bool

In [27]:
df_sessions.isna().any()

session_id                  False
client_id                   False
visit_date                  False
visit_time                  False
visit_number                False
utm_source                  False
utm_medium                  False
utm_campaign                False
utm_adcontent               False
device_category             False
device_brand                False
device_screen_resolution    False
device_browser              False
geo_country                 False
geo_city                    False
dtype: bool

2.3. Объединение таблиц.

Объединение таблиц выполняется по общему полю session_id. Однако для объединения не используется метод merge, поскольку не видится целесообразнмы получение таблицы большого объема с избыточностью. Вместо этого планируется проход по таблице sessions с переносом в нее фич только со значимыми аттрибутами. Т. е. выполняется проверка на наличие "целевого действия" в сессии. И если "да", то сохраняется наименьший номер hit_number в сессии, при котором было достигнуто "целевое действие" и сопутствующие ему данные. Если "нет", то наибольшее значение hit_number в сессии и также сопутствующие ему данные. В основе данного подхода лежит предположение, что чем дольше клиент взаимодействует с сервисом, тем больше вероятность достижения "целевого действия". И наоборот, клиенты не заинтересованные покидают сервис раньше.

In [33]:
list(df_sessions.session_id)

['9055434745589932991.1637753792.1637753792',
 '905544597018549464.1636867290.1636867290',
 '9055446045651783499.1640648526.1640648526',
 '9055447046360770272.1622255328.1622255328',
 '9055447046360770272.1622255345.1622255345',
 '9055447192389856083.1622453074.1622453074',
 '9055455318486370642.1640843788.1640843788',
 '9055461992850812764.1626107740.1626107740',
 '9055462349345527315.1638536723.1638536723',
 '9055466554104774132.1624800757.1624800757',
 '9055466554104774132.1629442326.1629442326',
 '9055469620715506713.1628883994.1628883994',
 '9055469620715506713.1633110583.1633110583',
 '9055469620715506713.1635878177.1635878177',
 '9055469620715506713.1636559415.1636559415',
 '9055476398179250925.1634235280.1634235280',
 '9055487268745225369.1637983385.1637983385',
 '9055487955932503299.1630493885.1630493885',
 '9055488359664698205.1635763037.1635763037',
 '9055491722626447422.1638119489.1638119489',
 '9055496369776380607.1633438303.1633438303',
 '9055497958914887309.1634045644.16

In [41]:
for element in list(df_sessions.session_id):
    print(df_hits[(df_hits.session_id == element) & (df_hits.event_action.isin([
        'sub_car_claim_click', 'sub_car_claim_submit_click', 'sub_open_dialog_click', 'sub_custom_question_submit_click',
        'sub_call_number_click', 'sub_callback_submit_click', 'sub_submit_success', 'sub_car_request_submit_click']))])
    break

Empty DataFrame
Columns: [session_id, hit_number, hit_page_path, event_category, event_action]
Index: []
